### PR TITLE
add websocket and code refactor

### PR DESCRIPTION
**Pull request by huojiao2006**

### CODE REVIEW

File: pkg/models/task.go [hunk 1]
```diff
@@ -62,6 +64,7 @@ func NewTaskId() string {
 }
 
 type TaskDirective struct {
+	NotificationId     string
```
Comment (chilianyi) [inline pkg/models/task.go:68]: There is notificationId in Task, is it needed here?

File: pkg/services/notification/resource_control/notification.go [hunk 2]
```diff
@@ -140,6 +150,7 @@ func SplitNotificationIntoTasks(ctx context.Context, notification *models.Notifi
 			task := models.NewTask(
 				notification.NotificationId,
 				jsonutil.ToString(directive),
+				address.NotifyType,
```
Comment (chilianyi) [inline pkg/services/notification/resource_control/notification.go]: notifyType already exists in directive, is it needed here?

File: pkg/services/notification/server.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ package notification
 import (
 	"google.golang.org/grpc"
 
+	"openpitrix.io/logger"
```
Comment (chilianyi) [inline pkg/services/notification/server.go:9]: need blank line

File: pkg/services/notification/validation.go [hunk 4]
```diff
@@ -50,6 +52,67 @@ func ValidateCreateNotificationParams(ctx context.Context, req *pb.CreateNotific
...
+						logger.Errorf(ctx, "Failed to validate email [%s]: %+v", email, err)
+						return err
+					}
+				}
+			}
+		}
+	}
+
+	//Step2: check addressInfo format is like address_info = ["adl-xxxx1", "adl-xxxx2"]
+	if err != nil {
```
Comment (chilianyi) [inline pkg/services/notification/validation.go]: elif will be better

File: api/notification.proto [hunk 5]
```diff
@@ -414,4 +427,17 @@ service notification {
 	}
 
 
+	rpc CreateWebsocketMessage (CreateWebsocketMessageRequest) returns (CreateWebsocketMessageResponse) {
```
Comment (chilianyi) [inline api/notification.proto]: Can this func merged into CreateNotification?

File: pkg/db/common.go [hunk 6]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/jinzhu/gorm"
```
Comment (chilianyi) [inline pkg/db/common.go]: no need empty line

File: pkg/db/db_test.go [hunk 7]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/jinzhu/gorm"
```
Comment (chilianyi) [inline pkg/db/db_test.go]: ditto

File: pkg/models/address_info.go [hunk 8]
```diff
@@ -30,3 +31,14 @@ func DecodeAddressListIds(data string) (*AddressListIds, error) {
 	}
 	return addressListIds, err
 }
+
+type NotificationExtra map[string]string
```
Comment (chilianyi) [inline pkg/models/address_info.go]: Mv this to pkg/models/notification.go

File: pkg/services/notification/resource_control/notification.go [hunk 9]
```diff
@@ -161,14 +212,83 @@ func SplitNotificationIntoTasks(ctx context.Context, notification *models.Notifi
...
+		}
+		logger.Infof(ctx, "Push user [%s] websocket message id [%s] ,Directive[%s] to etcd .", userId, task.TaskId, task.Directive)
+
+	}
+	return nil
+
+}
+
+func checkExtra(ctx context.Context, notification *models.Notification) error {
+	extrastr := notification.Extra
```
Comment (chilianyi) [inline pkg/services/notification/resource_control/notification.go]: change to extraStr

File: pkg/constants/common.go [hunk 10]
```diff
@@ -14,7 +14,7 @@ const (
 
 const (
 	EtcdPrefix                  = "notification/"
-	NotificationTaskTopicPrefix = "nf-task"
+	NotificationTaskTopicPrefix = "nf-email-task"
```
Comment (chilianyi) [inline pkg/constants/common.go]: need to change this prefix?

File: pkg/services/notification/validation.go [hunk 11]
```diff
@@ -53,6 +94,79 @@ func ValidateCreateNotificationParams(ctx context.Context, req *pb.CreateNotific
...
+				for _, email := range values {
+					err := VerifyEmailFmt(ctx, email)
+					if err != nil {
+						logger.Errorf(ctx, "Failed to validate email [%s]: %+v", email, err)
+						return err
+					}
+				}
+			}
+			if notifyType == constants.NotifyTypeWebsocket {
+				extrastr := notification.Extra
```
Comment (chilianyi) [inline pkg/services/notification/validation.go]: Already implements in checkExtra?

File: pkg/models/task.go (hunk 1 above)
Comment (huojiao2006) [reply pkg/models/task.go:68]: yes,will use to write socket message.

File: pkg/models/address_info.go (hunk 8 above)
Comment (huojiao2006) [reply pkg/models/address_info.go]: fix

File: pkg/services/notification/resource_control/notification.go (hunk 9 above)
Comment (huojiao2006) [reply pkg/services/notification/resource_control/notification.go]: fix

File: pkg/constants/common.go (hunk 10 above)
Comment (huojiao2006) [reply pkg/constants/common.go]: currently no need.

File: pkg/services/notification/resource_control/notification.go (hunk 2 above)
Comment (huojiao2006) [reply pkg/services/notification/resource_control/notification.go]: yes, need use it to distinguish different task if we add other type such as sms,wechat in the future.

File: pkg/services/notification/server.go (hunk 3 above)
Comment (huojiao2006) [reply pkg/services/notification/server.go:9]: fix.

File: api/notification.proto (hunk 5 above)
Comment (huojiao2006) [reply api/notification.proto]: yes

File: pkg/services/notification/validation.go (hunk 4 above)
Comment (huojiao2006) [reply pkg/services/notification/validation.go]: fix

File: pkg/db/common.go (hunk 6 above)
Comment (huojiao2006) [reply pkg/db/common.go]: fix

File: pkg/db/db_test.go (hunk 7 above)
Comment (huojiao2006) [reply pkg/db/db_test.go]: fix

File: pkg/models/notification.go [hunk 12]
```diff
@@ -121,3 +137,35 @@ func NotificationSet2PbSet(inNfs []*Notification) []*pb.Notification {
...
+	if err != nil {
+		logger.Errorf(nil, "Decode [%s] into notification extra failed: %+v", data, err)
+	}
+	return extra, err
+}
+
+func CheckExtra(ctx context.Context, notification *Notification) error {
+	extraStr := notification.Extra
+	if extraStr == "" {
+		logger.Errorf(ctx, "Failed to validate addressInfo, extra is blank: [%s].", extraStr)
```
Comment (chilianyi) [inline pkg/models/notification.go:155]: The log is not exact.
logger.Errorf(ctx, "Failed to validate notification extra [%s], should be: {\"ws_message_type\": \"xxx\"}", extraStr)
So do the logs below.

File: pkg/services/notification/handler.go [hunk 13]
```diff
@@ -146,7 +151,9 @@ func (s *Server) createTasks(ctx context.Context, tasks []*models.Task) error {
 		}
 		logger.Debugf(ctx, "Create task [%s] in DB successfully.", task.TaskId)
 
-		err = s.controller.taskQueue.Enqueue(task.TaskId)
+		if task.NotifyType == constants.NotifyTypeEmail {
+			err = s.controller.taskQueue.Enqueue(task.TaskId)
+		}
```
Comment (chilianyi) [inline pkg/services/notification/handler.go:161]: Mv the "}" after line 161

File: pkg/services/notification/validation.go [hunk 14]
```diff
@@ -53,6 +94,67 @@ func ValidateCreateNotificationParams(ctx context.Context, req *pb.CreateNotific
...
+		if len(*addressListIds) == 0 {
+			logger.Errorf(ctx, "Failed to validate addressInfo, address list id is blank: %+v", err)
+			return gerr.New(ctx, gerr.InvalidArgument, gerr.ErrorIllegalNotificationAddressList, notification.AddressInfo)
+		}
+
+		for _, addressListId := range *addressListIds {
+			//check addresslist id start with adl-, address_info = ["adl-xxxx1", "adl-xxxx2"]
+			prefix := string([]byte(addressListId)[:4])
+			if prefix != models.AddressListIdPrefix {
+				logger.Errorf(ctx, "Failed to validate addressInfo, address list id is invalid [%s]: %+v", addressListId, err)
```
Comment (chilianyi) [inline pkg/services/notification/validation.go]: the err is nil

File: pkg/services/notification/validation.go [hunk 15]
```diff
@@ -53,6 +94,67 @@ func ValidateCreateNotificationParams(ctx context.Context, req *pb.CreateNotific
...
+				return gerr.New(ctx, gerr.InvalidArgument, gerr.ErrorIllegalNotificationAddressList, notification.AddressInfo)
+			}
+		}
+	} else {
+		logger.Errorf(ctx, "Failed to validate addressInfo, decode address list id failed [%s]: %+v", notification.AddressInfo, err)
```
Comment (chilianyi) [inline pkg/services/notification/validation.go]: notification.AddressInfo not id.

File: pkg/services/notification/handler.go (hunk 13 above)
Comment (huojiao2006) [reply pkg/services/notification/handler.go:161]: fix

File: pkg/services/notification/validation.go (hunk 14 above)
Comment (huojiao2006) [reply pkg/services/notification/validation.go]: fix

File: pkg/services/notification/validation.go (hunk 15 above)
Comment (huojiao2006) [reply pkg/services/notification/validation.go]: fix

**Review by chilianyi — APPROVED** [review]

**Pull request merged**